### PR TITLE
[Infineon] Lock-app updates, adding multiple users and credentials for CYW30739

### PR DESCRIPTION
**Pull request by AlvinHsiao**
[Infineon] Lock-app updates, adding multiple users and credentials for CYW30739

    * Fix mCredentials, fix NVM flash saving

    * restyle

    * cleanup

    * remove unused array

    * remove TotalCredentials key

    * Check mMaxUsers against DOOR_LOCK_MAX_USERS

### CODE REVIEW

File: examples/lock-app/cyw30739/src/LockManager.cpp [hunk 1]
```diff
@@ -29,16 +30,28 @@ LockManager LockManager::sLock;
...
-    // Allocate buffer for credentials
-    if (!mCredentials.Alloc(maxNumberOfCredentialsPerUser))
+    for (uint8_t i = 0; i < ArraySize(mLockUsers); i++)
     {
-        printf("Failed to allocate array for lock credentials");
-        return CHIP_ERROR_NO_MEMORY;
+        // Allocate buffer for credentials
+        if (!mCredentials[i].Alloc(maxNumberOfCredentialsPerUser))
+        {
+            printf("Failed to allocate array for lock credentials");
```
Comment (tcarmelveilleux) [inline examples/lock-app/cyw30739/src/LockManager.cpp:40]: Suggest using ChipLogXxx to get timestamp/module/etc

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
Fast tracking platform changes, pending comments.

File: examples/lock-app/cyw30739/src/LockManager.cpp [hunk 2]
```diff
@@ -29,16 +30,28 @@ LockManager LockManager::sLock;
...
+            return APP_ERROR_ALLOCATION_FAILED;
+        }
     }
+
     mMaxCredentialsPerUser = maxNumberOfCredentialsPerUser;
 
+    mMaxUsers = numberOfSupportedUsers;
+    if (mMaxUsers > DOOR_LOCK_MAX_USERS)
```
Comment (bzbarsky-apple) [inline examples/lock-app/cyw30739/src/LockManager.cpp:48]: Should this be using `ArraySize(mLockUsers)`?

File: examples/lock-app/cyw30739/src/LockManager.cpp [hunk 3]
```diff
@@ -205,67 +221,76 @@ bool LockManager::Unlock(chip::EndpointId endpointId, const Optional<chip::ByteS
...
     if (DlUserStatus::kAvailable == user.userStatus)
     {
-        ChipLogDetail(Zcl, "Found unoccupied user [endpoint=%d]", mEndpointId);
+        ChipLogDetail(Zcl, "Found unoccupied user [endpoint=%d]", endpointId);
         return true;
     }
 
     user.userName       = chip::CharSpan(userInDb.userName.data(), userInDb.userName.size());
-    user.credentials    = chip::Span<const DlCredential>(userInDb.credentials.data(), userInDb.credentials.size());
+    user.credentials    = chip::Span<const DlCredential>(mCredentials[adjustedUserIndex].Get(), userInDb.credentials.size());
```
Comment (bzbarsky-apple) [inline examples/lock-app/cyw30739/src/LockManager.cpp:241]: Why is this change needed?

File: examples/lock-app/cyw30739/src/LockManager.cpp [hunk 4]
```diff
@@ -275,58 +300,77 @@ bool LockManager::SetUser(uint16_t userIndex, chip::FabricIndex creator, chip::F
...
     // Save user information in NVM flash
-    CYW30739Config::WriteConfigValueBin(CYW30739Config::kConfigKey_LockUser, reinterpret_cast<const uint8_t *>(&userInStorage),
-                                        sizeof(EmberAfPluginDoorLockUserInfo));
+    CYW30739Config::WriteConfigValueBin(CYW30739Config::kConfigKey_LockUser, reinterpret_cast<const uint8_t *>(&mLockUsers),
+                                     sizeof(EmberAfPluginDoorLockUserInfo) * mMaxUsers);
 
     CYW30739Config::WriteConfigValueBin(CYW30739Config::kConfigKey_UserCredentials,
-                                        reinterpret_cast<const uint8_t *>(mCredentials.Get()), sizeof(DlCredential));
+                                     reinterpret_cast<const uint8_t *>(mCredentials[adjustedUserIndex].Get()),
+                                     sizeof(DlCredential) * totalCredentials);
```
Comment (bzbarsky-apple) [inline examples/lock-app/cyw30739/src/LockManager.cpp:316]: This does not make sense to me.  This is writing to the single kConfigKey_UserCredentials key, but writing out the credentials for one user?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: examples/lock-app/cyw30739/src/LockManager.cpp [hunk 5]
```diff
@@ -275,58 +300,77 @@ bool LockManager::SetUser(uint16_t userIndex, chip::FabricIndex creator, chip::F
...
     // Save user information in NVM flash
-    CYW30739Config::WriteConfigValueBin(CYW30739Config::kConfigKey_LockUser, reinterpret_cast<const uint8_t *>(&userInStorage),
-                                        sizeof(EmberAfPluginDoorLockUserInfo));
+    CYW30739Config::WriteConfigValueBin(CYW30739Config::kConfigKey_LockUser, reinterpret_cast<const uint8_t *>(&mLockUsers),
+                                     sizeof(EmberAfPluginDoorLockUserInfo) * mMaxUsers);
 
     CYW30739Config::WriteConfigValueBin(CYW30739Config::kConfigKey_UserCredentials,
-                                        reinterpret_cast<const uint8_t *>(mCredentials.Get()), sizeof(DlCredential));
+                                     reinterpret_cast<const uint8_t *>(mCredentials[adjustedUserIndex].Get()),
+                                     sizeof(DlCredential) * mMaxCredentialsPerUser);
```
Comment (bzbarsky-apple) [inline examples/lock-app/cyw30739/src/LockManager.cpp:316]: This is writing out _one user's credentials_, no?  It's writing a buffer of size `sizeof(DlCredential) * mMaxCredentialsPerUser)`.

The corresponding read later will read `sizeof(DlCredential) * mMaxUsers * mMaxCredentialsPerUser` bytes and will read into a totally different place in memory.

Can we please stop the guess-and-check, and in the next update respond to this comment with a clear description of how storage is meant to work that I can then compare to the code?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: examples/lock-app/cyw30739/src/LockManager.cpp (hunk 5 above)
Comment (AlvinHsiao) [reply examples/lock-app/cyw30739/src/LockManager.cpp]: Sorry, I referred to https://github.com/project-chip/connectedhomeip/pull/20044 and updated a new commit.
It should write out all user's credentials.
